redis/asyncio/connection.py

```diff
@@ -922,55 +881,27 @@ async def can_read(self, timeout: float = 0):
                 f"Error while reading from {self.host}:{self.port}: {e.args}"
             )
 
-    async def read_response(self, disable_decoding: bool = False):
-        """Read the response from a previously sent command"""
-        try:
-            async with self._lock:
-                if self.socket_timeout:
-                    async with async_timeout.timeout(self.socket_timeout):
-                        response = await self._parser.read_response(
-                            disable_decoding=disable_decoding
-                        )
-                else:
-                    response = await self._parser.read_response(
-                        disable_decoding=disable_decoding
-                    )
-        except asyncio.TimeoutError:
-            await self.disconnect()
-            raise TimeoutError(f"Timeout reading from {self.host}:{self.port}")
-        except OSError as e:
-            await self.disconnect()
-            raise ConnectionError(
-                f"Error while reading from {self.host}:{self.port} : {e.args}"
-            )
-        except BaseException:
-            await self.disconnect()
-            raise
-
-        if self.health_check_interval:
-            if sys.version_info[0:2] == (3, 6):
-                func = asyncio.get_event_loop
-            else:
-                func = asyncio.get_running_loop
-            self.next_health_check = func().time() + self.health_check_interval
-
-        if isinstance(response, ResponseError):
-            raise response from None
-        return response
-
-    async def read_response_without_lock(self, disable_decoding: bool = False):
+    async def read_response(
+        self,
+        disable_decoding: bool = False,
+        timeout: Optional[float] = None,
+        with_lock: bool = True,
+    ):
         """Read the response from a previously sent command"""
+        read_timeout = timeout if timeout is not None else self.socket_timeout
+        lock_ctxt = self._lock if with_lock else nullcontext
+        if not self.is_connected:
+            await self.connect()
         try:
-            if self.socket_timeout:
-                async with async_timeout.timeout(self.socket_timeout):
-                    response = await self._parser.read_response(
-                        disable_decoding=disable_decoding
-                    )
-            else:
+            async with lock_ctxt, async_timeout.timeout(read_timeout):
```

Comment:
Wouldn't an if/else on with_lock directly be more performant instead of a null context? Same for async_timeout. Even with 0 timeout, async_timeout adds significant latency.

Reply:
That is a matter of opinion.  pub-sub is primarily IO bound.  An application spends its time *waiting* for a message.  Once a message arrives, two extra calls to __aexit__() make no difference in the large scheme of things.

The `not with_lock` is a special case, used only in cluster mode.
the async_timeout is usually needed, and always for timeout=0.  It could be skipped for `timeout==None` but this blocking case is, again, an uncommon special case.

There is an argument for reducing code duplication and simplifying program flow to ensure correctness, even if the cost is an extra __aexit__ call in special circumstances.    The alternative is to have *four* differenc control paths, for `with_lock == True/False` and `timeout is None or not None`.  In my opinion, that is not very pythonic.

Reply:
Normally I would always argue against code duplication, but the foremost feature of redis is its speed. If we've to make small ergonomic sacrifices so that the code is faster, then I think it should be considered.
Also, async calls add significantly more overhead than sync calls, which is why I even bothered commenting. Out of the two, we should at least have an if/else for the timeout block, as it adds a lot of latency.
And if you are keen on removing the if/else for _without_lock function, can you at least directly call this new function instead of adding a level of indirection in cluster?

Reply:
I'll oblige with the changes you suggest, of course, because getting fixes into async redis-py has been a very arduous process indeed.

But as a python developer for 20 years, allow me to suggest that the changes will not amount to any measurable difference.  Remember, this is Python we're talking about.  Redis may be fast, but all that is dwarfed by the fact that we are running inside Python.  We already have layers and layers of buffers, abstractions, and classes, even lower than this function of which you speak. An extra null function call, in some minority special case, just does not matter.

My suggestion would be, instead, to simplify the code as much as possible removing interdependencies and special cases, so that in further steps, we can remove those intermediate buffering layers.  The async code in redis-py suffers too much from being verbatim copy of the blocking code, with the odd `async` keyword thrown in.

Reply:
I didn't add the direct call to the without_lock function in cluster since I didn't want the changes to spread too far, nor change the api of the connection.py too be backwards incompatible.  But removing the special _without_lock function seems very reasonable.

(why the without-lock is required in the cluster client is also not clear to me, not documented anywhere, maybe you can explain why there is no need for locking in the cluster client?  I've never used cluster)

Reply:
Btw, maybe you are confused about `async_timeout.timeout()`.  If `timeout==None`, then this is just a null context manager, and the result is a *blocking* operator, with negligible overhead added.   If `timeout==0` then this amounts to *polling* the connection.  the timeout context manager doesn't add any IO latency as such, any *latency* it adds comes from the cost of a context manager and the cost of calling Python functions.

Btw, I once advocated that the Python interpreter accept `None` as a null context manager, effectively omitting it, making the special `contextlib.nullcontext()` unnecessary, but it was rejected by the core devs:
https://bugs.python.org/issue18677

Reply:
Async Cluster implementation is not using the connection pool from asyncio client implementation, which uses locks for providing multi-process & multi-threading support. Async Cluster doesn't support multi-process yet & multi-threading isn't supported with asyncio anyway: https://docs.python.org/3/library/asyncio-dev.html#asyncio-multithreading.

It also uses deques instead of lists/sets, so adding/removing connections from the pool is already atomic. These differences improved speed by 9-10% AFAIR. The deque change should be adopted for standalone async clients as well.

We can keep this function for backwards compatibility, but can you please change the functions in cluster client?

Reply:
async_timeout doesn't return nullcontext if timeout is none: https://github.com/aio-libs/async-timeout/blob/master/async_timeout/__init__.py#L119

Reply:
right not an actual nullcontext, that was just a turn of phrase, but it is still a no-op.  if `timeout is None`, it does nothing.

Anyway, I´ve made the fixes you suggested, permuting the context manager code.  I also changed the calls to the _without_lock from the Cluster code and tests.

Which actually got me thinking, locking (async) really shouldn't be done anyway.  It should be up to the client to perform locking if there is more than one `Task` reading from a connection at the same time, which is not necessarily a good architecture anyway.  The connection pool should perform any necessary locking to be async-safe, since it is designed to be used by many clients.

How about removing the lock alltogether, making sure that an underlying connection pool is safe to use by multiple Tasks?  I'll look the code over, and see if it makes sense.

Reply:
Yes, the `with_lock` is just wrapped around the parser `read_response`.  it doesn't even cover any ConnectionPool code.  Unless I'm mistaken, we can eliminate the lock completely.
Also, the use of the lock in the ConnectionPool can be removed.  Locks are only needed in async code if there are any (potentially blocking) async calls going on.  The connection management can be made completely lock-free if desired.

I can wrap more such changes into this PR, or make a a separate one for removing asyncio locking.  What do you think?

Reply:
The connectionpool can be simplified.  It is already not thread-safe.  (since async IO and threading doesn't mix.  And having it Fork-safe is also dubious.

Reply:
I also suggest that the fork-protection be removed from asyncio connection pools.  When forking using asyncio, all bets are off.  Threading and forking are fundamentally incompatible with asyncio.  I´ll make a separate PR for that.

Reply:
I would say a separate PR would probably be better, since it may need other changes and/or improved tests.

I agree, threading support should neither be supported nor be expected.
Regarding removing multiprocessing support, I'm definitely in favour of it. But @chayim @dvora-h would've to take a call for release plan.

Reply:
I prefer tight, surgical PRs. It allows us to both review code, comment effectively - and frankly, helps us not miss things. This includes surgical tests.

Feature parity (where possible) belongs in all of our usage models:
1. Sync
2. Async
3. Sync Cluster
4. Async Cluster

I'm generally against removing features that people are using. However, I'm pro pre-announced deprecation (see the Python announcement in the README).